lib/puppet/http/resolver/serverlist.rb

```diff
@@ -0,0 +1,30 @@
+class Puppet::HTTP::Resolver::ServerList < Puppet::HTTP::Resolver
+  def initialize(client, server_list: Puppet[:server_list], default_port: Puppet[:masterport])
+    @client = client
+    @server_list = server_list
+    @default_port = default_port
+  end
+
+  def resolve(session, name, &block)
+    success = false
+    @server_list.each do |server|
+      host = server[0]
+      port = server[1] || @default_port
+      begin
+        response = @client.get(URI("https://#{host}:#{port}/status/v1/simple/master"))
+        if response.success?
+          success = true
+          yield session.create_service(name, host, port)
```

Comment:
@joshcooper here we're yielding every successful connection to the session. If I'm understanding correctly, we'll then have multiple services running that the session can then select? I'm not sure I fully understand how this is beneficial

Reply:
My thinking is that all resolvers should yield candidate services, one at a time, and let the session decide whether to use that service, or continue looking. This way the "resolution policy" is in the session. And future changes in the policy only need to happen in the session. If the resolver starts making policy decisions, then it gets more complicated to determine what the policy is, and how to change it.

Currently the resolution policy is hardcoded in the session as "walk each resolver and use the first service that we can connect to". But in the future, we might want a different policy, such as "and also, make sure it supports the version 5 of the catalog API endpoint".